### PR TITLE
fix(36055): `prop in dict` does not report error if `prop` is union type such as `MyClass | string`

### LINKED ISSUE #36055: `prop in dict` does not report error if `prop` is union type such as `MyClass | string`
**TypeScript Version:**  3.8.0-dev.20200104

**Search Terms:**
is:issue is:open in union in:title 
[This was hard to search for since "in" is such a common word]

**Code**

```ts
class Skipped { }

declare const skipped: Skipped;
declare const skippedOrString: Skipped | string;

const dict = { "a": 1, "b": 2, "c": 3 };

console.log("a" in dict);
console.log(skipped in dict);
console.log(skippedOrString in dict);
```

**Expected behavior:**

Should get a compile error on the last line (as well as the line before that), since `skippedOrString` is of type `Skipped | string`, where `Skipped` is a class, and thus can't be a dictionary key.

**Actual behavior:**

The second-to-last line does report an error, as it should; but the last line does not.

**Playground Link:**

http://www.typescriptlang.org/play/index.html?ts=3.8.0-dev.20200104&ssl=1&ssc=1&pln=11&pc=1#code/MYGwhgzhAEDKDWBLADsgpgE2gb2gXwCgCM1QwAnNaYAewDsIAXaCJVTALjjfQwG5ipcJWr0mLHpgDy5WI3KI6Acy4IUvaAB8W8xUoEFaDZhkTBmAXhzQARGBtcAjABpbAIwfQATK5vBPAMz4BkYQNCBoAHQgNEoAFHY20IrQpuYAlAKh4VEx8azqmMl0qWaMmYZiOdGxcQXsGDJyCsrFpRkCQA

**Related Issues:**

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #36055

### CODE REVIEW

**Comment by sandersn** [comment]
Reviewer list is broken for me, but @elibarzilay can you take a look at this too?

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -27797,7 +27797,8 @@ namespace ts {
             // The in operator requires the left operand to be of type Any, the String primitive type, or the Number primitive type,
             // and the right operand to be of type Any, an object type, or a type parameter type.
             // The result is always of the Boolean primitive type.
-            if (!(isTypeComparableTo(leftType, stringType) || isTypeAssignableToKind(leftType, TypeFlags.NumberLike | TypeFlags.ESSymbolLike))) {
+            if (!allTypesAssignableToKind(leftType, TypeFlags.StringLike | TypeFlags.NumberLike | TypeFlags.ESSymbolLike) &&
```
Comment (elibarzilay) [inline src/compiler/checker.ts:27800]: Two comments:

1. I had to mentally convert your `(!A && !(B && C))` to `(!(A || (B && C)))` to follow the conditions, so I think that the latter is more readable.
2. Is there a need for the `isTypeComparableTo` test?  I tried dropping it, and passed the tests.

```suggestion
            if (!(allTypesAssignableToKind(leftType, TypeFlags.StringLike | TypeFlags.NumberLike | TypeFlags.ESSymbolLike) ||
                  isTypeAssignableToKind(leftType, TypeFlags.Index | TypeFlags.TypeParameter))) {
```

Comment (a-tarasyuk) [reply src/compiler/checker.ts:27800]: @elibarzilay Thanks for the feedback. Fixed.

**Comment by sandersn** [comment]
@DanielRosenwasser this is a pretty nice fix, but technically a breaking change since it adds errors. I'm going to hold it for 4.0 unless you say otherwise since it's for a backlog bug. (Actually, just merge it if you want it in 3.9.)

**Pull request merged**